Shell/Options.cpp

```diff
@@ -1019,418 +1068,423 @@ void Options::init()
-            _termAlgebraCyclicityCheck.description=
-              "Activates the cyclicity rule for term algebras (such as algebraic datatypes in SMT-LIB):\n"
-              "- off : the cyclicity rule is not enforced (this is sound but incomplete)\n"
-              "- axiom : the cyclicity rule is axiomatized with a transitive predicate describing the subterm relation over terms\n"
-              "- rule : the cyclicity rule is enforced by a specific hyper-resolution rule\n"
-              "- light : the cyclicity rule is enforced by rule generating disequality between a term and its known subterms";
-            _lookup.insert(&_termAlgebraCyclicityCheck);
-            _termAlgebraCyclicityCheck.tag(OptionTag::INFERENCES);
-
-	    _forwardDemodulation = ChoiceOptionValue<Demodulation>("forward_demodulation","fd",Demodulation::ALL,{"all","off","preordered"});
-	    _forwardDemodulation.description=
-	    "Oriented rewriting of newly derived clauses by kept unit equalities\n"
-	    "s = t     L[sθ] \\/ C\n"
-	    "---------------------  where sθ > tθ\n"
-	    " L[tθ] \\/ C\n"
-	    "If 'preordered' is set, only equalities s = t where s > t are used for rewriting.";
-	    _lookup.insert(&_forwardDemodulation);
-	    _forwardDemodulation.tag(OptionTag::INFERENCES);
-	    _forwardDemodulation.setRandomChoices({"all","all","all","off","preordered"});
+    _unificationWithAbstraction = ChoiceOptionValue<UnificationWithAbstraction>("unification_with_abstraction","uwa",
+                                     UnificationWithAbstraction::OFF,
+                                     {"off","interpreted_only","one_side_interpreted","one_side_constant","all","ground"});
+    _unificationWithAbstraction.description=
+      "During unification, if two terms s and t fail to unify we will introduce a constraint s!=t and carry on. For example, "
+      "resolving p(1) \\/ C with ~p(a+2) would produce C \\/ 1 !=a+2. This is controlled by a check on the terms. The expected "
+      "use case is in theory reasoning. The possible values are:"
+      "- off: do not introduce a constraint\n"
+      "- interpreted_only: only if s and t have interpreted top symbols\n"
+      "- one_side_interpreted: only if one of s or t have interpreted top symbols\n"
+      "- one_side_constant: only if one of s or t is an interpreted constant (e.g. a number)\n"
+      "- all: always apply\n"
+      "- ground: only if both s and t are ground\n"
+      "See Unification with Abstraction and Theory Instantiation in Saturation-Based Reasoning for further details.";
+    _unificationWithAbstraction.tag(OptionTag::INFERENCES);
+    _lookup.insert(&_unificationWithAbstraction);
+
+    _useACeval = BoolOptionValue("use_ac_eval","uace",true);
+    _useACeval.description="Evaluate associative and commutative operators e.g. + and *.";
+    _useACeval.tag(OptionTag::INFERENCES);
```

Comment:
Add problem constraint for  hasTheories?

Reply:
Yeah I think that's a shout. For this one as well as for 
- _inequalityNormalization
- _unificationWithAbstraction
- _cancellation
- _highSchool
- _gaussianVariableElimination
- _pushUnaryMinus
- _arithmeticSubtermGeneralizations